src/Agent/NewRelic/Agent/Core/Transactions/Transaction.cs

```diff
@@ -1180,19 +1180,30 @@ public void InsertDistributedTraceHeaders<T>(T carrier, Action<T, string, string
             }
         }
 
-        public ITransaction SetRequestHeaders(IEnumerable<KeyValuePair<string, string>> parameters)
+        public ITransaction SetRequestHeaders<T>(T headers, IEnumerable<string> keysToCapture, Func<T, string, string> getter)
         {
-            if (parameters == null)
+            if (headers == null)
             {
-                throw new ArgumentNullException(nameof(parameters));
+                throw new ArgumentNullException(nameof(headers));
             }
 
-            foreach (var parameter in parameters)
+            if (keysToCapture == null)
             {
-                if (parameter.Key != null && parameter.Value != null)
+                throw new ArgumentNullException(nameof(keysToCapture));
+            }
+
+            if (getter == null)
+            {
+                throw new ArgumentNullException(nameof(getter));
+            }
+
+            foreach (var key in keysToCapture)
+            {
+                var value = getter(headers, key);
+                if (!string.IsNullOrEmpty(value))
```

Comment:
I think I recall seeing something in the spec about capturing empty parameter values, so we might want to only do a null check here.

Reply:
you are right. [spec](https://source.datanerd.us/agents/agent-specs/blob/32153bcabe8389ef03d0c7b18371bbccc72c7989/Agent-Attributes-PORTED.md#nullempty-attribute-values)